net: lwm2m: Added commands to the lwm2m_client shell

otkulseng opened the pull request:
- Added read, write and execute commands to the shell.

- Added start, stop and update to the shell.

- Refactored the event_cb of the rd_client_info struct into the ctx, as it was needed in the shell script.

Signed-off-by: Ola Tangen Kulseng <ola.kulseng@nordicsemi.no>

Commit:
net: lwm2m: Added commands to the lwm2m_client shell

Added read, write and execute commands to the shell.

Signed-off-by: Ola Tangen Kulseng <ola.kulseng@nordicsemi.no>

stephanosio reviewed (commented):
Please review the [Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines) and format the commit message accordingly.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -71,9 +89,192 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+	int ret = lwm2m_engine_get_resource(pathstr, &res);
+	if (ret < 0)
+	{
+		shell_error(sh, "Couldn't execute resource\n");
+		return ret;
```
I believe you should not return just any error code. See https://docs.zephyrproject.org/latest/doxygen/html/group__shell__api.html#ga2546bd682966f71f03985f220ff48dc2

So maybe just print out the error code, and then return either `-ENOEXEC` or `-EINVAL`
```suggestion
		shell_error(sh, "Resource not found (err %d)\n", ret);
		return -EINVAL;
```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -71,9 +89,192 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+		shell_error(sh, "Couldn't execute resource\n");
+		return ret;
+	}
+	if(res->execute_cb)
+	{
+		uint16_t obj_id;
+		sscanf(pathstr, "%*d/%hd/%*d", &obj_id);
+		return res->execute_cb(obj_id, argv[ignore_cnt], argc-ignore_cnt);
+	}
+	return 0;
```
Now you don't print out anything, if the resource does not have execute callback.
Also, we cannot return what the callback returned, because from shell point of view, the execution was success. So maybe print it out, if the return was not zero.
```suggestion
	if(!res->execute_cb) {
	        shell_error(sh, "No execute callback\n");
	        return -EINVAL;
	}        
	uint16_t obj_id;

	sscanf(pathstr, "%*d/%hd/%*d", &obj_id);
	ret = res->execute_cb(obj_id, argv[ignore_cnt], argc-ignore_cnt);
	if (ret) {
	        shell_error(sh, "returned %d\n", ret);
	        return -ENOEXEC;
	}
	return 0;
```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -71,9 +89,192 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+	int ret = 0;
+	
+	if (strcmp(dtype, "-s") == 0){
+		const char* buff;
+		uint16_t buff_len=0;
+		uint8_t buff_flags;
+		ret = lwm2m_engine_get_res_data(pathstr, (void **)&buff, &buff_len, &buff_flags);
```
Zephyr coding style requires you to have one empty line between variable introduction and the next line of code.

```suggestion
		const char* buff;
		uint16_t buff_len=0;
		uint8_t buff_flags;

		ret = lwm2m_engine_get_res_data(pathstr, (void **)&buff, &buff_len, &buff_flags);
```

This this similarly to rest of the cases as well.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -71,9 +89,192 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+		bool temp; 
+		ret = lwm2m_engine_get_bool(pathstr, &temp);
+		shell_print(sh, "%d\n", temp);
+	}else {
```
```suggestion
	} else {
```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -71,9 +89,192 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+	}
+	
+	if (ret < 0) {
+		shell_error(sh, "can't do read operation, request failed\n");
+		return ret;
```
```suggestion
		shell_error(sh, "can't do read operation, request failed (err %d)\n", ret);
		return -ENOEXEC;
```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -71,9 +89,192 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+	}
+
+	const char* value = argv[3];
+	char* e;
+	if (strcmp(dtype, "-s") == 0){
```
```suggestion
	char* e;

	if (strcmp(dtype, "-s") == 0){
```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -71,9 +89,192 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+		return -EINVAL;
+	}
+
+	int ret = 0;
+	const char* pathstr = argv[1];
+
+	const char* dtype = "-s"; /* default */
+	if (argc == 4) /* write path options value */
```
```suggestion
	int ret = 0;
	const char* pathstr = argv[1];
	const char* dtype = "-s"; /* default */

	if (argc == 4) /* write path options value */
```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -71,9 +89,192 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+		return -EINVAL;
+	}
+
+	if (*e != '\0')
+	{
```
```suggestion
	if (*e != '\0') {
```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -71,9 +89,192 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+	} else if (strcmp(dtype, "-f") == 0){
+		double temp=strtod(value, &e);
+		ret = lwm2m_engine_set_float(pathstr, &temp);
+	}else {
```
```suggestion
	} else {
```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -71,9 +89,192 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+	char* e;
+	if (strcmp(dtype, "-s") == 0){
+		ret = lwm2m_engine_set_string(pathstr, argv[2]);
+	} else if (strcmp(dtype, "-s8") == 0){
```
```suggestion
	} else if (strcmp(dtype, "-s8") == 0) {
```
Fix this on all rest of these as well. Space is missing before the opening bracket.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -71,9 +89,192 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+	const char* pathstr = argv[1];
+	int ret = 0;
+	
+	if (strcmp(dtype, "-s") == 0){
```
```suggestion
	if (strcmp(dtype, "-s") == 0) {
```
Space missing before the opening bracket. Fix this for all clauses in this function as well.

SeppoTakalo reviewed (changes requested):
Looking good.
I left few remarks to be fixed. Otherwise functionally looks OK.

VeijoPesonen commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -11,16 +11,37 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #include <zephyr/kernel.h>
 #include <zephyr/net/lwm2m.h>
 #include <zephyr/shell/shell.h>
+<<<<<<< HEAD
```
I see you have a small issue here - I assume you tried to rebase your changes but some artifacts were left which should be removed.

VeijoPesonen commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -11,16 +11,34 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+	"Paths are inserted without leading '/'\n"
+#define LWM2M_HELP_READ "Read value from LwM2M resource\nread [PATH] [OPTIONS]...\n" \
+	"-s \tRead value as string(default)\n" \
+	"-b \tRead value as boole (1/0)\n" \
```
```suggestion
	"-b \tRead value as bool (1/0)\n" \
```

rlubos commented:
Please check compliance results, there are some errors reported. Also, the code doesn't seem to build according to test results.

SeppoTakalo commented:
Ah.. there is one more thing to modify, some build warnings:
```
[1/6] Building C object zephyr/subsys/net/lib/lwm2m/CMakeFiles/subsys__net__lib__lwm2m.dir/lwm2m_shell.c.obj
/home/seta/src/ncs/zephyr/subsys/net/lib/lwm2m/lwm2m_shell.c: In function ‘cmd_read’:
/home/seta/src/ncs/zephyr/subsys/net/lib/lwm2m/lwm2m_shell.c:151:3: warning: ‘lwm2m_engine_get_res_data’ is deprecated [-Wdeprecated-declarations]
  151 |   ret = lwm2m_engine_get_res_data(pathstr, (void **)&buff, &buff_len, &buff_flags);
      |   ^~~
In file included from /home/seta/src/ncs/zephyr/subsys/net/lib/lwm2m/lwm2m_shell.c:18:
/home/seta/src/ncs/zephyr/include/zephyr/net/lwm2m.h:1091:5: note: declared here
 1091 | int lwm2m_engine_get_res_data(const char *pathstr, void **data_ptr, uint16_t *data_len,
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~

```

Commit:
net: lwm2m: Add LwM2M shell commands for start&stop

Added start, stop and update to the shell.
Refactored the event_cb of the rd_client_info struct into the ctx,
as it was needed in the shell script.

Signed-off-by: Ola Tangen Kulseng <ola.kulseng@nordicsemi.no>

SeppoTakalo reviewed (dismissed):
Looks good to me.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -11,16 +11,32 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define LWM2M_HELP_CMD "LwM2M commands"
 #define LWM2M_HELP_SEND "LwM2M SEND operation\nsend [OPTION]... [PATH]...\n" \
 	"-n\tSend as non-confirmable\n" \
-	"Paths are inserted without leading '/'\n" \
```
I didn't get answer for my question. If the documentation here was wrong, this change should be reflected in the commit message.

SeppoTakalo replied:
The documentation was wrong. Paths worked with or without the leading slash.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -67,13 +90,307 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+		if (ret == 0) {
+			ret = lwm2m_engine_set_res_data_len(pathstr, buff_len + 1);
+		}
+		shell_print(sh, "%s\n", buff);
```
Shouldn't all those prints here and below only take place if the operation suceeds? Oterwise, the command will end up printing some rubbish in case of failure.

otkulseng replied:
I tested the other commands, and it did not seem to matter whether or not the paths were inserted without leading '/'. Haven't tested the send cmd however, so reverted my change.

otkulseng replied:
> Shouldn't all those prints here and below only take place if the operation suceeds? Oterwise, the command will end up printing some rubbish in case of failure.

Yes. Fixed it now.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -67,13 +90,307 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+		uint16_t buff_len = 0;
+
+		ret = lwm2m_engine_get_res_buf(pathstr, (void **)&buff, &buff_len, NULL, NULL);
+		if (ret == 0) {
+			ret = lwm2m_engine_set_res_data_len(pathstr, buff_len + 1);
+		}
```
```suggestion
		if (ret != 0) {
			goto out;
		}
```

Oops.. I only now just spotted the error here. I might have been misguiding the author.
The `set_read_data_len()` is only required if we modify the buffer. In this case we are not.

Add a common error handling for all of these cases. Otherwise we end up printing some garbage, like Robert indicated earlier.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_shell.c:
```diff
@@ -62,18 +86,367 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
+		if (ret != 0) {
+			goto out;
+		}
+		ret = lwm2m_engine_set_res_data_len(pathstr, buff_len + 1);
+		if (ret != 0) {
+			goto out;
+		}
```
```suggestion
```
No need to touch the buffer length at all.
We are not modifying here, just reading.
